### PR TITLE
gradle: refactor 

### PR DESCRIPTION
**Pull request by volhovm**
Added gradle25 support, fixed installation paths "gradle/lib => lib/gradle/lib"
cc @jagajaga

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -5648,7 +5648,9 @@ let
 
   gotty = goPackages.gotty.bin // { outputs = [ "bin" ]; };
 
-  gradle = callPackage ../development/tools/build-managers/gradle { };
+  gradleGen = callPackage ../development/tools/build-managers/gradle { };
+  gradle = gradleGen.gradle28;
+  gradle25 = gradleGen.gradle25;
```
Comment (nbp) [inline pkgs/top-level/all-packages.nix]: Use `self.gradleGen.gradle25` instead, as this might later be a security issue.

**Comment by volhovM** [comment]
Added those self.PREV attributes.

**Comment by nbp** [comment]
cc @MarcWeber @iyzsong @bbarker @bjornfor @dochang Anybody interested in doing this review, and also becoming a maintainer for this package?

**Comment by jagajaga** [comment]
:+1:

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -5648,7 +5648,9 @@ let
 
   gotty = goPackages.gotty.bin // { outputs = [ "bin" ]; };
 
-  gradle = callPackage ../development/tools/build-managers/gradle { };
+  gradleGen = callPackage ../development/tools/build-managers/gradle { };
+  gradle = self.gradleGen.gradle28;
+  gradle25 = self.gradleGen.gradle25;
```
Comment (danbst) [inline pkgs/top-level/all-packages.nix]: there is line
```
gradle25 = callPackage ../development/tools/build-managers/gradle/2.5.nix { };
```
right below. shouldn't it be removed?

You can also use `callPackages` idiom, as in https://github.com/NixOS/nixpkgs/pull/9834

**Comment by volhovM** [comment]
Fixed double declaration

**Comment by bendlas** [comment]
Please reopen and merge my fix at ee42c84
Has anybody even tested this, before filing the PR and/or merging?

@jagajaga I know you like to merge a lot & that's cool, but you know that, as a reviewer, you're also supposed to build & test run stuff locally before merging?

**Comment by jagajaga** [comment]
@bendlas I do **always** build nontrivial stuff before merging.

**Comment by volhovM** [comment]
@bendlas That was my fault. Local builds were OK, dunno why I didn't noticed that. 
@jagajaga merged PR because I asked him personally and assured him everything worked.

**Comment by bendlas** [comment]
OK, guys, sorry for getting worked up. It _was_ an easy-to-miss error, since a cursory test did show the java help screen instead of a proper error message, which might not have been apparent to a non-regular java user.

So, if somebody could confirm, that with my commit gradle runs again, we can get this fixed, no?

**Comment by jagajaga** [comment]
What test exactly should we run? (I'm not a java guy myself :) )

**Comment by bendlas** [comment]
A cursory test would be to just run `gradle` and see if it's actually gradle that has been run and you're not seeing the java help screen. You can see that by looking for the gradle version string.
If you want to add a unit test for gradle, you could try to replicate the following shell session:
```sh
$ git clone https://github.com/spring-guides/gs-gradle.git
$ cd gs-gradle/initial
$ gradle build
# ignore gradle output
$ java -cp build/classes/main hello.HelloWorld
Hello World! # look for this output
```
When you see that last `Hello World!` output string, you know that gradle has successfully compiled spring's getting-started project, so that java could run it.

**Comment by bendlas** [comment]
@jagajaga any luck with the review? gradle is still broken in master, where it has worked before ...

**Comment by jagajaga** [comment]
Can you create a PR with your changes? And I will test it with your example.

**Comment by bendlas** [comment]
Changes are attached to this ticket, see above. Here is my formal pull request:
```
Please pull commit ee42c84 in the branch gradle-fix-refactor from my repository,
it contains a fix for the gradle package, which has been broken as part of
previous work on this ticket.
```
Here is a convenient command line to do this: `git fetch https://github.com/bendlas/nixpkgs.git gradle-fix-refactor && git cherry-pick ee42c84`

**Comment by jagajaga** [comment]
@bendlas oh, sorry. Ok, will test now.

**Comment by jagajaga** [comment]
@bendlas That works, thank you.

**Comment by jagajaga** [comment]
https://github.com/NixOS/nixpkgs/commit/b3b7f9f37f8c134aa22b66ec6a0a1d555f1638bb

**Comment by bendlas** [comment]
thanks

**Pull request merged**